File: examples/nodejs/send-trace.js
#!/usr/bin/env node
/**
 * Example: Send OTLP traces to zradar from Node.js
 * 
 * This demonstrates how to:
 * 1. Configure OpenTelemetry SDK for Node.js
 * 2. Authenticate with API key
 * 3. Send traces to zradar
 * 4. Instrument LLM calls
 */

import { trace } from '@opentelemetry/api';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';

// Configuration
const ZVRADAR_ENDPOINT = process.env.ZVRADAR_ENDPOINT || 'localhost:4317';
const API_KEY = process.env.ZVRADAR_API_KEY || 'your-api-key-here';

/**
 * Setup OpenTelemetry SDK
 */
function setupTelemetry() {
  // Create resource with service information
  const resource = new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: 'example-nodejs-app',
    [SemanticResourceAttributes.SERVICE_VERSION]: '1.0.0',
    [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: 'development',
  });

  // Configure OTLP exporter with authentication
  const traceExporter = new OTLPTraceExporter({
    url: `http://${ZVRADAR_ENDPOINT}`,
    headers: {
      authorization: `Bearer ${API_KEY}`,
    },
  });

  // Create and configure SDK
  const sdk = new NodeSDK({
    resource,
    traceExporter,
  });

  // Start the SDK
  sdk.start();
  console.log(`✅ Telemetry configured - sending to ${ZVRADAR_ENDPOINT}`);

  // Graceful shutdown
  process.on('SIGTERM', () => {
    sdk.shutdown()
      .then(() => console.log('Telemetry terminated'))
      .catch((error) => console.log('Error terminating telemetry', error))
      .finally(() => process.exit(0));
  });

  return trace.getTracer('zradar-example');
}

/**
 * Simulate an OpenAI API call
 */
async function simulateOpenAICall(tracer) {
  const span = tracer.startSpan('llm.openai.completion');
  
  try {
    // Set LLM-specific attributes
    span.setAttributes({
      'llm.vendor': 'openai',
      'llm.model': 'gpt-4',
      'llm.temperature': 0.7,
      'llm.max_tokens': 1000,
      'llm.stream': false,
    });

    // Simulate API latency
    await new Promise(resolve => setTimeout(resolve, 500));

    // Add response metadata
    span.setAttributes({
      'llm.prompt.tokens': 50,
      'llm.completion.tokens': 150,
      'llm.total.tokens': 200,
      'llm.cost.input': 0.0015,
      'llm.cost.output': 0.0045,
      'llm.cost.total': 0.006,
      'llm.finish_reason': 'stop',
    });

    console.log('✅ Simulated OpenAI call');
  } catch (error) {
    span.recordException(error);
    span.setStatus({ code: 2, message: error.message }); // ERROR
    throw error;
  } finally {
    span.end();
  }
}

/**
 * Simulate Anthropic Claude API call
 */
async function simulateClaudeCall(tracer) {
  const span = tracer.startSpan('llm.anthropic.completion');
  
  try {
    span.setAttributes({
      'llm.vendor': 'anthropic',
      'llm.model': 'claude-3-opus-20240229',
      'llm.temperature': 0.8,
      'llm.max_tokens': 2000,
      'llm.top_p': 0.95,
    });

    await new Promise(resolve => setTimeout(resolve, 600));

    span.setAttributes({
      'llm.prompt.tokens': 75,
      'llm.completion.tokens': 200,
      'llm.total.tokens': 275,
      'llm.cost.total': 0.021, // Claude Opus pricing
      'llm.stop_reason': 'end_turn',
    });

    console.log('✅ Simulated Claude call');
  } finally {
    span.end();
  }
}

/**
 * Simulate a RAG (Retrieval Augmented Generation) pipeline
 */
async function simulateRAGPipeline(tracer) {
  const rootSpan = tracer.startSpan('pipeline.rag');
  
  try {
    rootSpan.setAttributes({
      'pipeline.type': 'rag',
      'pipeline.query': 'What is zradar?',
    });

    // Step 1: Generate query embedding
    const embedSpan = tracer.startSpan('step.generate_embedding', {}, trace.setSpan(trace.context.active(), rootSpan));
    try {
      embedSpan.setAttributes({
        'llm.vendor': 'openai',
        'llm.model': 'text-embedding-ada-002',
        'embedding.dimensions': 1536,
        'llm.cost.total': 0.0001,
      });
      await new Promise(resolve => setTimeout(resolve, 150));
    } finally {
      embedSpan.end();
    }

    // Step 2: Vector database search
    const searchSpan = tracer.startSpan('step.vector_search', {}, trace.setSpan(trace.context.active(), rootSpan));
    try {
      searchSpan.setAttributes({
        'db.system': 'pinecone',
        'db.operation': 'query',
        'results.count': 5,
        'results.relevance.threshold': 0.8,
      });
      await new Promise(resolve => setTimeout(resolve, 100));
    } finally {
      searchSpan.end();
    }

    // Step 3: LLM completion with context
    const completionSpan = tracer.startSpan('step.completion', {}, trace.setSpan(trace.context.active(), rootSpan));
    try {
      completionSpan.setAttributes({
        'llm.vendor': 'openai',
        'llm.model': 'gpt-4',
        'context.chunks': 5,
        'llm.total.tokens': 800,
        'llm.cost.total': 0.024,
      });
      await new Promise(resolve => setTimeout(resolve, 700));
    } finally {
      completionSpan.end();
    }

    rootSpan.setAttributes({
      'pipeline.success': true,
      'pipeline.total_cost': 0.0241,
    });

    console.log('✅ Simulated RAG pipeline');
  } finally {
    rootSpan.end();
  }
}

/**
 * Simulate error handling
 */
async function simulateErrorScenario(tracer) {
  const span = tracer.startSpan('llm.call.with_retry');
  
  try {
    span.setAttributes({
      'llm.vendor': 'openai',
      'llm.model': 'gpt-4',
    });

    // Simulate retries
    for (let attempt = 1; attempt <= 3; attempt++) {
      const retrySpan = tracer.startSpan(`attempt.${attempt}`, {}, trace.setSpan(trace.context.active(), span));
      
      try {
        retrySpan.setAttribute('retry.attempt', attempt);
        
        if (attempt < 3) {
          // Simulate failure
          await new Promise(resolve => setTimeout(resolve, 100));
          const error = new Error('Rate limit exceeded');
          retrySpan.recordException(error);
          retrySpan.setAttribute('error', true);
          retrySpan.setAttribute('error.type', 'RateLimitError');
        } else {
          // Success on third attempt
          await new Promise(resolve => setTimeout(resolve, 500));
          retrySpan.setAttribute('success', true);
        }
      } finally {
        retrySpan.end();
      }
    }

    span.setAttributes({
      'retry.total_attempts': 3,
      'retry.success': true,
    });

    console.log('✅ Simulated retry scenario');
  } finally {
    span.end();
  }
}

/**
 * Main function
 */
async function main() {
  console.log('🚀 zradar OTLP Example - Node.js\n');

  // Check for API key
  if (API_KEY === 'your-api-key-here') {
    console.log('⚠️  Warning: Using placeholder API key');
    console.log('   Set ZVRADAR_API_KEY environment variable with your actual key');
    console.log('   Example: export ZVRADAR_API_KEY=zvr_abc123...\n');
  }

  // Setup telemetry
  const tracer = setupTelemetry();

  console.log('\nSending example traces...\n');

  try {
    // Example 1: OpenAI call
    console.log('1. OpenAI GPT-4 completion:');
    await simulateOpenAICall(tracer);

    // Example 2: Anthropic Claude call
    console.log('\n2. Anthropic Claude completion:');
    await simulateClaudeCall(tracer);

    // Example 3: RAG pipeline
    console.log('\n3. RAG pipeline:');
    await simulateRAGPipeline(tracer);

    // Example 4: Error handling with retries
    console.log('\n4. Error handling with retries:');
    await simulateErrorScenario(tracer);

    console.log('\n⏳ Flushing spans to zradar...');
    
    // Give time for spans to be exported
    await new Promise(resolve => setTimeout(resolve, 2000));

    console.log('\n✨ Done! Check zradar dashboard to see your traces');
    console.log('   Admin API: http://localhost:8081');
    console.log('   Swagger UI: http://localhost:8081/swagger-ui/');

  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }

  process.exit(0);
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

File: examples/python/send_trace.py
#!/usr/bin/env python3
"""
Example: Send OTLP traces to zradar

This demonstrates how to:
1. Configure OpenTelemetry SDK
2. Authenticate with API key
3. Send traces to zradar
4. Add custom attributes for LLM observability
"""

import time
import os
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.resources import Resource
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter

# Configuration
ZVRADAR_ENDPOINT = os.getenv("ZVRADAR_ENDPOINT", "localhost:4317")
API_KEY = os.getenv("ZVRADAR_API_KEY", "your-api-key-here")

def setup_telemetry():
    """Configure OpenTelemetry to send traces to zradar"""
    
    # Create resource with service information
    resource = Resource.create({
        "service.name": "example-llm-app",
        "service.version": "1.0.0",
        "deployment.environment": "development",
    })
    
    # Create tracer provider
    provider = TracerProvider(resource=resource)
    
    # Configure OTLP exporter with authentication
    exporter = OTLPSpanExporter(
        endpoint=ZVRADAR_ENDPOINT,
        insecure=True,  # Use False with TLS in production
        headers=(
            ("authorization", f"Bearer {API_KEY}"),
        ),
    )
    
    # Add batch processor
    processor = BatchSpanProcessor(exporter)
    provider.add_span_processor(processor)
    
    # Set as global tracer provider
    trace.set_tracer_provider(provider)
    
    print(f"✅ Telemetry configured - sending to {ZVRADAR_ENDPOINT}")
    return trace.get_tracer(__name__)


def simulate_llm_call(tracer):
    """Simulate an LLM API call with rich instrumentation"""
    
    with tracer.start_as_current_span("llm.completion") as span:
        # Standard OTLP attributes
        span.set_attribute("gen_ai.system", "openai")
        span.set_attribute("gen_ai.request.model", "gpt-4")
        span.set_attribute("gen_ai.request.temperature", 0.7)
        span.set_attribute("gen_ai.request.max_tokens", 1000)
        
        # Custom LLM-specific attributes
        span.set_attribute("gen_ai.usage.input_tokens", 45)
        span.set_attribute("gen_ai.usage.output_tokens", 123)
        span.set_attribute("gen_ai.usage.total_tokens", 168)
        
        # Cost tracking
        span.set_attribute("llm.cost.prompt_usd", 0.0014)  # $0.0014
        span.set_attribute("llm.cost.completion_usd", 0.0037)  # $0.0037
        span.set_attribute("llm.cost.total_usd", 0.0051)
        
        # Request/response metadata
        span.set_attribute("gen_ai.response.id", "resp_456def")
        span.set_attribute("gen_ai.response.finish_reasons", "stop")
        
        # Simulate processing time
        time.sleep(0.5)
        
        # Nested span for prompt processing
        with tracer.start_as_current_span("llm.prompt.format") as prompt_span:
            prompt_span.set_attribute("prompt.template", "chat_completion")
            prompt_span.set_attribute("prompt.variables", 3)
            time.sleep(0.1)
        
        # Simulate response processing
        with tracer.start_as_current_span("llm.response.parse") as response_span:
            response_span.set_attribute("response.format", "json")
            response_span.set_attribute("response.valid", True)
            time.sleep(0.05)
        
        print("✅ Simulated LLM call")


def simulate_embedding_call(tracer):
    """Simulate an embedding generation call"""
    
    with tracer.start_as_current_span("llm.embedding") as span:
        span.set_attribute("gen_ai.system", "openai")
        span.set_attribute("gen_ai.request.model", "text-embedding-ada-002")
        span.set_attribute("embedding.dimensions", 1536)
        span.set_attribute("embedding.input.tokens", 20)
        span.set_attribute("llm.cost.total_usd", 0.0002)
        
        time.sleep(0.2)
        print("✅ Simulated embedding call")


def simulate_error_scenario(tracer):
    """Simulate an error case with proper instrumentation"""
    
    with tracer.start_as_current_span("llm.completion.failed") as span:
        span.set_attribute("gen_ai.system", "openai")
        span.set_attribute("gen_ai.request.model", "gpt-4")
        
        try:
            # Simulate an error
            raise Exception("Rate limit exceeded")
        except Exception as e:
            span.set_attribute("error", True)
            span.set_attribute("error.type", type(e).__name__)
            span.set_attribute("error.message", str(e))
            span.record_exception(e)
            print("✅ Simulated error scenario")


def simulate_complex_workflow(tracer):
    """Simulate a complex multi-step LLM workflow"""
    
    with tracer.start_as_current_span("workflow.document_qa") as root_span:
        root_span.set_attribute("workflow.type", "question_answering")
        root_span.set_attribute("workflow.document.id", "doc_789")
        
        # Step 1: Generate embeddings for chunks
        with tracer.start_as_current_span("step.embed_chunks") as embed_span:
            embed_span.set_attribute("chunks.count", 5)
            for i in range(5):
                with tracer.start_as_current_span(f"chunk.embed.{i}") as chunk_span:
                    chunk_span.set_attribute("chunk.index", i)
                    chunk_span.set_attribute("chunk.tokens", 100)
                    time.sleep(0.05)
        
        # Step 2: Query vector database
        with tracer.start_as_current_span("step.vector_search") as search_span:
            search_span.set_attribute("query.embedding.dimensions", 1536)
            search_span.set_attribute("results.count", 3)
            search_span.set_attribute("results.relevance.min", 0.85)
            time.sleep(0.1)
        
        # Step 3: Generate answer
        with tracer.start_as_current_span("step.generate_answer") as answer_span:
            answer_span.set_attribute("gen_ai.system", "openai")
            answer_span.set_attribute("gen_ai.request.model", "gpt-4")
            answer_span.set_attribute("context.chunks", 3)
            answer_span.set_attribute("gen_ai.usage.total_tokens", 500)
            answer_span.set_attribute("llm.cost.total_usd", 0.015)
            time.sleep(0.3)
        
        root_span.set_attribute("workflow.duration_ms", 500)
        root_span.set_attribute("workflow.success", True)
        print("✅ Simulated complex workflow")


def main():
    """Main example runner"""
    print("🚀 zradar OTLP Example - Python\n")
    
    # Check for API key
    if API_KEY == "your-api-key-here":
        print("⚠️  Warning: Using placeholder API key")
        print("   Set ZVRADAR_API_KEY environment variable with your actual key")
        print("   Example: export ZVRADAR_API_KEY=zvr_abc123...\n")
    
    # Setup telemetry
    tracer = setup_telemetry()
    
    print("\nSending example traces...\n")
    
    # Example 1: Simple LLM call
    print("1. Simple LLM completion:")
    simulate_llm_call(tracer)
    
    # Example 2: Embedding generation
    print("\n2. Embedding generation:")
    simulate_embedding_call(tracer)
    
    # Example 3: Error scenario
    print("\n3. Error scenario:")
    simulate_error_scenario(tracer)
    
    # Example 4: Complex workflow
    print("\n4. Complex multi-step workflow:")
    simulate_complex_workflow(tracer)
    
    # Flush all pending spans
    print("\n⏳ Flushing spans to zradar...")
    trace.get_tracer_provider().force_flush()
    
    print("\n✨ Done! Check zradar dashboard to see your traces")
    print(f"   Admin API: http://localhost:8081")
    print(f"   Swagger UI: http://localhost:8081/swagger-ui/")


if __name__ == "__main__":
    main()

